### PR TITLE
HPCC-20169 Spark configuration and runtime

### PR DESCRIPTION
**Pull request by Michael-Gardner**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@xwang2713 Please review the changes made to your packaging code.

**Review by xwang2713 — COMMENTED** [review]
Looks good.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt [hunk 1]
```diff
@@ -37,18 +37,24 @@ set(dfuserver "dfuserver" "--daemon %i" "dafilesrv.service")
 set(eclagent "agentexec" "--daemon %i" "dafilesrv.service")
 set(eclccserver "eclccserver" "--daemon %i" "dafilesrv.service")
 set(eclscheduler "eclscheduler" "--daemon %i" "dafilesrv.service")
-set(esp "esp" "--daemon %i" "dafilesrv.service")
```
Comment (richardkchapman) [inline initfiles/bash/etc/systemd/system/CMakeLists.txt]: Why was this line deleted? Seems unrelated (and may explain the smoketest failure?)

File: initfiles/bash/etc/systemd/system/component@instance.service.in [hunk 2]
```diff
@@ -7,7 +7,7 @@ Requires=hpcc-conf@%i.service
 Type=forking
 User=@RUNTIME_USER@
 Group=@RUNTIME_GROUP@
-PermissionsStartOnly=true
+PermissionsStartOnly=false
```
Comment (richardkchapman) [inline initfiles/bash/etc/systemd/system/component@instance.service.in]: Is this related? I've seen this change in a different PR...

File: initfiles/bash/etc/systemd/system/esp@instance.service.in [hunk 3]
```diff
@@ -0,0 +1,27 @@
+[Unit]
```
Comment (richardkchapman) [inline initfiles/bash/etc/systemd/system/esp@instance.service.in]: Is this related? Not sure how it is connected with Spark

File: initfiles/bash/etc/systemd/system/thorslave@instance.service.in [hunk 4]
```diff
@@ -6,10 +6,12 @@ After=network.target remote-fs.target @dependencies@
 Type=simple
 User=@RUNTIME_USER@
 Group=@RUNTIME_GROUP@
-PermissionsStartOnly=true
+PermissionsStartOnly=false
```
Comment (richardkchapman) [inline initfiles/bash/etc/systemd/system/thorslave@instance.service.in]: is this related?

**Comment by richardkchapman** [comment]
@Michael-Gardner Some concerns about changes that seem unrelated, and the smoketest failure

**Comment by xwang2713** [comment]
@Michael-Gardner if "-DUSE_SPARK" turn off which I believe the smoke test does do we skip Spark configuration and startup when HPCC started?

**Comment by Michael-Gardner** [comment]
Looking into the smoke test failure

File: initfiles/bash/etc/systemd/system/thorslave@instance.service.in (hunk 4 above)
Comment (Michael-Gardner) [reply initfiles/bash/etc/systemd/system/thorslave@instance.service.in]: This one should be in here.  The one in another PR is regarding the thor@instance.service.in file I believe, but I'll double check.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt (hunk 1 above)
Comment (Michael-Gardner) [reply initfiles/bash/etc/systemd/system/CMakeLists.txt]: This line was deleted because there is now an esp@instance.service file, and we don't generate it from the component@instance.service template.

File: initfiles/bash/etc/systemd/system/esp@instance.service.in (hunk 3 above)
Comment (Michael-Gardner) [reply initfiles/bash/etc/systemd/system/esp@instance.service.in]: This file is needed because we now start spark along with the esp.  That's why we broke it out into it's own separate file instead of having it get built by a template.

File: spark/spark-defaults.conf.in [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+# Example:
+# spark.master                     spark://master:7077
+# spark.eventLog.enabled           true
+# spark.eventLog.dir               hdfs://namenode:8021/directory
+# spark.serializer                 org.apache.spark.serializer.KryoSerializer
+# spark.driver.memory              5g
+# spark.executor.extraJavaOptions	-Xint
+
+# Path for native libraries
+spark.driver.extraLibraryPath @LIB_PATH@
```
Comment (rpastrana) [inline spark/spark-defaults.conf.in]: where is LIB_PATH set? what is it set to?

File: spark/spark-env.sh.in [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+# - SPARK_PID_DIR       Where the pid file is stored. (Default: /tmp)
+# - SPARK_IDENT_STRING  A string representing this instance of spark. (Default: $USER)
+# - SPARK_NICENESS      The scheduling priority for daemons. (Default: 0)
+# - SPARK_NO_DAEMONIZE  Run the proposed command in the foreground. It will not output a PID file.
+# Options for native BLAS, like Intel MKL, OpenBLAS, and so on.
+# You might get better performance to enable these options if using native BLAS (see SPARK-21305).
+# - MKL_NUM_THREADS=1        Disable multi-threading of Intel MKL
+# - OPENBLAS_NUM_THREADS=1   Disable multi-threading of OpenBLAS
+
+ALL_IPS=$(/opt/HPCCSystems/sbin/configgen -env /etc/HPCCSystems/environment.xml -listall2 | awk -F , '{print $3}' | sort | uniq)
```
Comment (rpastrana) [inline spark/spark-env.sh.in]: can these paths be generated based on the CMAKE vars we set up for install location, etc?

**Review by rpastrana — APPROVED** [review]
@Michael-Gardner just a couple of comments/questions. 
Also, as Richard pointed out, there seems to be some overlap w/ another commit, do we need a rebase?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@Michael-Gardner as is, the startup of ESP/Thor will be affected by the startup of spark, we should consider a independent hpcc-spark startup component.

**Comment by JamesDeFabia** [comment]
@g-pan Please review the docID tags

File: initfiles/componentfiles/configxml/spark.xsd [hunk 7]
```diff
@@ -20,88 +20,204 @@
...
-    <xs:key name="thorProcessKey1">
-      <xs:selector xpath="./ThorMasterProcess|./ThorSlaveProcess"/>
-      <xs:field xpath="@name"/>
-    </xs:key>
-  </xs:element>
+        <xs:complexType>
+            <!--DOC-Autobuild-code-->
+            <xs:annotation>
+                <xs:appinfo>
+                    <docid>Th.t1</docid>
```
Comment (g-pan) [inline initfiles/componentfiles/configxml/spark.xsd]: please change to <docid>Sprk.t1</docid>

**Review by g-pan — COMMENTED** [review]
please change the <docid> value in the spark.xsd file, if possible

File: initfiles/bash/etc/systemd/system/CMakeLists.txt [hunk 8]
```diff
@@ -29,7 +29,9 @@ set(componentList
     sasha
     thor
     thorslave
-    backupnode)
+    backupnode
+    spark
```
Comment (rpastrana) [inline initfiles/bash/etc/systemd/system/CMakeLists.txt]: why do we have two spark related components listed here, and should they be renamed?

File: initfiles/bin/init_sparkthor.in [hunk 9]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Comment (rpastrana) [inline initfiles/bin/init_sparkthor.in]: update date.

File: initfiles/componentfiles/configxml/cgencomplist_win.xml [hunk 10]
```diff
@@ -70,6 +70,9 @@
   <Component name="backupnode" processName="BackupNode" schema="backupnode.xsd">
     <File name="backupnode.xsl" method="xslt" destName="backupnode.xml"/>
   </Component>
+  <Component name="sparkthor" processName="SparkThor" schema="sparkThor.xsd">
+    <File name="sparkThor.xsl" method="xslt" destName="spark-env.conf"/>
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/cgencomplist_win.xml]: this entry differs between cgencomplist_win.xml and cgencomplist_linux.xml by the desName:
"spark-env.conf" vs "spark-env.sh" is this intentional?

File: initfiles/componentfiles/configxml/environment.xsd [hunk 11]
```diff
@@ -795,7 +795,7 @@
             <xs:field xpath="@program"/>
         </xs:keyref>
         <xs:key name="processKey">
-            <xs:selector xpath=".//EclServerProcess|.//EclAgentProcess|.//SybaseProcess|.//SpareProcess|.//HoleCluster|.//ThorCluster|.//DaliServer|.//BackupNodeProcess"/>
+		<xs:selector xpath=".//EclServerProcess|.//EclAgentProcess|.//SybaseProcess|.//SpareProcess|.//HoleCluster|.//ThorCluster|.//DaliServer|.//BackupNodeProcess|.//SparkThorProcess"/>
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/environment.xsd]: minor, but indet seems off

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 12]
```diff
@@ -0,0 +1,232 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: date

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+################################################################################
+-->
+
+<!-- edited with XMLSPY v5 rel. 4 U (http://www.xmlspy.com) by Rajeev Moondhra (Seisint, Inc.) -->
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: lets remove this copy/paste remnant

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
+    <xs:include schemaLocation="environment.xsd"/>
+    <xs:element name="SparkThorProcess">
+        <xs:annotation>
+            <xs:documentation>Describes a spark cluster</xs:documentation>
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: Describes a Spark-Thor cluster

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        </xs:annotation>
+        <xs:complexType>
+            <!--DOC-Autobuild-code-->
+            <xs:annotation>
+                <xs:appinfo>
+                    <docid>Th.t1</docid>
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: @g-pan please review

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+                </xs:appinfo>
+            </xs:annotation>  
+            <xs:sequence>
+                <xs:element name="Instance" maxOccurs="unbounded">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: It's hard to visualize, but if this is a container of instances, it needs to be removed (sparkthor will inherit its instances from the referenced thor cluster)

File: initfiles/componentfiles/configxml/sparkThor.xsl.in [hunk 17]
```diff
@@ -0,0 +1,113 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems®.
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsl.in]: date

File: initfiles/componentfiles/configxml/sparkThor.xsl.in [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+            <xsl:with-param name="var" select="'SPARK_WORKER_OPTS'"/>
+            <xsl:with-param name="val" select="@SPARK_WORKER_OPTS"/>
+        </xsl:call-template>
+        <xsl:call-template name="printVariable">
+            <xsl:with-param name="var" select="'SPARK_DAEMON_MEMORY'"/>
+            <xsl:with-param name="val" select="@SPARK_DAEMON_MEMORY"/>
+        </xsl:call-template>
+        <xsl:text>SPARK_PID_DIR=@PID_PATH@</xsl:text>
+        <xsl:text>SPARK_CONF_DIR=@RUNTIME_PATH@/</xsl:text><xsl:value-of select="@name"/>
+    </xsl:template>
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsl.in]: can we verify that NodeGroup/@name exists in the actual /Environment/Software/ThorCluster/@name ?

File: initfiles/sbin/spark-hpcc.sh.in [hunk 19]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (rpastrana) [inline initfiles/sbin/spark-hpcc.sh.in]: this file might need renaming to sparkthor.sh.in

File: initfiles/sbin/spark-hpcc.sh.in [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+export SPARK_LOG_DIR="@LOG_PATH@/${_component}"
+export SPARK_PID_DIR="@PID_PATH@"
+if [[ "${LOCAL_IP}" == "${MASTER_IP}" ]]; then
+    @INSTALL_DIR@/externals/spark-hadoop/sbin/${_command}-master.sh
+fi
+
+export SPARK_WORKER_DIR="@RUNTIME_PATH@/${_component}"
+if [[ ${_IS_THORSLAVE} == 0 ]]; then
+    @INSTALL_DIR@/externals/spark-hadoop/sbin/${_command}-slave.sh ${MASTER_URL}
```
Comment (rpastrana) [inline initfiles/sbin/spark-hpcc.sh.in]: is spark-hadoop a folder we create, or is it created by the spark install process?

File: initfiles/sbin/spark-hpcc.sh.in [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+
+_command=${1:-start}
+_component=${2:-myspark-hpcc}
+
+LOCAL_IP="$(@ADMIN_PATH@/configgen -env @CONFIG_DIR@/@ENV_XML_FILE@ -listall -ip localhost | awk -F "," 'NR==1{print $3;}')"
+@ADMIN_PATH@/configgen -env @CONFIG_DIR@/@ENV_XML_FILE@ -listall2 -ip localhost -t thor | grep "ThorSlaveProcess" 2>&1 1>/dev/null
+_IS_THORSLAVE=$?
+# Always use the first record returned by `configgen -t esp` as master
+MASTER_IP="$(@ADMIN_PATH@/configgen -env @CONFIG_DIR@/@ENV_XML_FILE@ -listall -t esp | awk -F "," 'NR==1{print $3;}')"
+MASTER_URL="spark://${MASTER_IP}:7077"
```
Comment (rpastrana) [inline initfiles/sbin/spark-hpcc.sh.in]: is this port configurable?

File: spark/spark-defaults.conf.in [hunk 22]
```diff
@@ -0,0 +1,35 @@
+#
```
Comment (rpastrana) [inline spark/spark-defaults.conf.in]: @fvillanustre as part of the spark-hpcc integration effort, we'll be supplying spark config files dynamically configured. However those files include the following license. Are we ok to do this?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@Michael-Gardner a few comments. We need to add a sparkthor to the stock env.xml (when use_spark) used to build/package. Also, I don't see where we pull information from the newly created config to dynamically configure spark.

**Review by rpastrana — CHANGES_REQUESTED** [review]
@Michael-Gardner

File: initfiles/sbin/spark-hpcc.sh.in (hunk 20 above)
Comment (Michael-Gardner) [reply initfiles/sbin/spark-hpcc.sh.in]: It's a part of the spark install process.

File: initfiles/bin/init_sparkthor.in [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+source @INSTALL_DIR@/etc/init.d/hpcc_common
+set_environmentvars
+source @ADMIN_PATH@/hpcc_setenv ""
+
+component=$(basename ${PWD})
+logpthtail="$(date +%Y_%m_%d_%H_%M_%S)"
+export logfile="@LOG_PATH@/${component}/${component}_${logpthtail}.log"
+
+exec 3>&2 2>/var/log/HPCCSystems/${component}.debug
```
Comment (rpastrana) [inline initfiles/bin/init_sparkthor.in]: "/var/log/HPCCSystems" should be variable (@LOG_PATH@ perhaps) ?

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+                </xs:annotation>
+            </xs:attribute>
+            <xs:attribute name="SPARK_EXECUTOR_MEMORY" type="xs:nonNegativeInteger" default="1G">
+                <xs:annotation>
+                    <xs:appinfo>
+                        <tooltip>Memory per executor</tooltip>
+                    </xs:appinfo>
+                </xs:annotation>
+            </xs:attribute>
+            <xs:attribute name="SPARK_MASTER_HOST">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: this should be a configmanager computer instance

File: initfiles/sbin/CMakeLists.txt [hunk 25]
```diff
@@ -37,6 +37,15 @@ ADD_CUSTOM_TARGET(ProcessFiles-initfiles-sbin ALL DEPENDS ${outFiles} ${noInstal
 configure_file("${CMAKE_CURRENT_SOURCE_DIR}/regex.awk.in.cmake" "${CMAKE_CURRENT_BINARY_DIR}/regex.awk")
 configure_file("${CMAKE_CURRENT_SOURCE_DIR}/backup.sh.in" "${CMAKE_CURRENT_BINARY_DIR}/backup.sh" @ONLY)
 
+if(USE_SPARK)
```
Comment (rpastrana) [inline initfiles/sbin/CMakeLists.txt]: we also need to configure HPCC-Platform/initfiles/etc/DIR_NAME/environment.xml.in
 if USE_SPARK, to add stock "mysparkthor" associated with "mythor" and sparkmaster="myesp/Instance"

File: initfiles/sbin/spark-hpcc-slave.sh.in [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+_component=${2:-mysparkthor}
+
+mkdir -p @RUNTIME_PATH@ @LOCK_PATH@ @PID_PATH@ @LOG_PATH@
+@ADMIN_PATH@/configgen -env @CONFIG_DIR@/@ENV_XML_FILE@ -c ${_component} -od @RUNTIME_PATH@ -id @INSTALL_DIR@/componentfiles/configxml 
+source ./spark-env.sh
+
+cd @RUNTIME_PATH@/${_component}
+
+MASTER_IP="$(@ADMIN_PATH@/configgen -env @CONFIG_DIR@/@ENV_XML_FILE@ -listall -t sparkthor | awk -F "," '{print $3;}')"
+MASTER_URL="spark://${MASTER_IP}:7077"
```
Comment (rpastrana) [inline initfiles/sbin/spark-hpcc-slave.sh.in]: is the spark master port configurable?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@Michael-Gardner a few comments.

**Comment by Michael-Gardner** [comment]
*Note  -- use daliadmin instead of configgen to build the worker list

File: spark/spark-defaults.conf.in (hunk 5 above)
Comment (Michael-Gardner) [reply spark/spark-defaults.conf.in]: generally it will be set to /opt/HPCCSystems/lib

File: initfiles/sbin/CMakeLists.txt (hunk 25 above)
Comment (Michael-Gardner) [reply initfiles/sbin/CMakeLists.txt]: It is now included in the default environment.xml

**Comment by Michael-Gardner** [comment]
@rpastrana 
- included in the default environment.xml.
- now uses daliadmin to build the workers list
- renamed from spark-hpcc to sparkthor for all relevant scripts
- fixed issue of workers not starting from correct directory  on remote machines

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 27]
```diff
@@ -108,52 +88,18 @@
...
+            <xs:attribute name="NodeGroup" type="xs:string" use="required" default="mythor">
                 <xs:annotation>
                     <xs:appinfo>
-                        <tooltip>Configuration directory for hadoop</tooltip>
+                        <tooltip>Thor group for workers to attach to</tooltip>
                     </xs:appinfo>
                 </xs:annotation>
             </xs:attribute>
-            <xs:attribute name="SPARK_LOCAL_IP" type="xs:nonNegativeInteger">
+            <xs:attribute name="daliServers" type="daliServersType" use="required">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: can this be inherited from the thor cluster config or from the overall environmen?

File: initfiles/componentfiles/configxml/sparkThor.xsd [hunk 28]
```diff
@@ -206,27 +145,13 @@
                     </xs:appinfo>
                 </xs:annotation>
             </xs:attribute>
-            <xs:attribute name="SPARK_WORKER_PORT" type="xs:nonNegativeInteger" default="7070">
+            <xs:attribute name="SPARK_WORKER_PORT" type="xs:nonNegativeInteger" default="7071">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/sparkThor.xsd]: this is b/c dali tends to run on 7070?

**Review by rpastrana — COMMENTED** [review]
@Michael-Gardner looks good, just a couple of comments

File: initfiles/componentfiles/configxml/sparkThor.xsd (hunk 28 above)
Comment (Michael-Gardner) [reply initfiles/componentfiles/configxml/sparkThor.xsd]: Yes.  And this gets written out to the spark-env.sh file so all the workers will be able to see the new default port.

**Comment by Michael-Gardner** [comment]
@rpastrana Squashed the commit and kept the history of all our changes.

**Review by rpastrana — APPROVED** [review]
@Michael-Gardner 
Looks good.
@jpmcmu and @gfortil tested

**Comment by Michael-Gardner** [comment]
@richardkchapman Ready for merge.
